generic/piranha-tree-sitter/src/piranha.rs

```diff
@@ -0,0 +1,534 @@
+      while let Some(parent) = current_node.parent() {
+        // Check if the parent matches the `rule.constraint.matcher`
+        // This is the scope in which the constraint query will be applied.
+        if let Some((range, _)) = parent.get_match_for_query(
+          &self.code,
+          &constraint.matcher().create_query(rule_store.language()),
+          false,
+        ) {
+          let scope_node = self.get_node_for_range(range.start_byte, range.end_byte);
+          // Apply each query within the `scope_node`
+          for query_with_holes in constraint.queries() {
+            let query_str = query_with_holes.substitute_tags(&capture_by_tags);
+            let query = &rule_store.get_query(&query_str);
+            // If this query matches anywhere within the scope, return false.
+            if scope_node
+              .get_match_for_query(&self.code, query, true)
+              .is_some()
+            {
+              return false;
+            }
+          }
+          break;
+        }
+        current_node = parent;
+      }
+    }
+    true
+  }
+
+  /// Replaces the given byte range (`replace_range`) with the `replacement`.
+  /// Returns tree-sitter's edit representation along with updated source code.
+  /// Note: This method does not update `self`.
+  pub fn get_edit(&self, replace_range: Range, replacement: &str) -> (String, InputEdit) {
+    let source_code = self.code.clone();
+    // Create the new source code content by appropriately
+    // replacing the range with the replacement string.
+    let new_source_code = [
+      &source_code[..replace_range.start_byte],
+      replacement,
+      &source_code[replace_range.end_byte..],
+    ]
+    .concat();
+
+    // Log the edit
+    let replaced_code_snippet = &source_code[replace_range.start_byte..replace_range.end_byte];
+    #[rustfmt::skip]
+        info!("{} at ({:?}) -\n {}", if replacement.is_empty() { "Delete code" } else {"Update code" }.green(),
+            ((&replace_range.start_point.row, &replace_range.start_point.column),
+                (&replace_range.end_point.row, &replace_range.end_point.column)),
+            if !replacement.is_empty() {format!("{}\n to \n{}",replaced_code_snippet.italic(),replacement.italic())
+            } else {format!("{} ", replaced_code_snippet.italic())}
+        );
+
+    // Inner function to find the position (Point) for a given offset.
+    fn position_for_offset(input: &Vec<u8>, offset: usize) -> Point {
+      let mut result = Point { row: 0, column: 0 };
+      for c in &input[0..offset] {
+        if *c as char == '\n' {
+          result.row += 1;
+          result.column = 0;
+        } else {
+          result.column += 1;
+        }
+      }
+      result
+    }
+
+    let len_new_source_code_bytes = replacement.as_bytes().len();
+    let byte_vec = &source_code.as_bytes().to_vec();
+    // Create the InputEdit as per the tree-sitter api documentation.
+    let edit = InputEdit {
```

Comment:
Can this throw an error if the `start_byte`, etc is malformed? If so, handling for that?

Reply:
What do you mean by malformed?